File: Matlab/demo/Demo2.m
%% DEMO2  Comparison of Full and Sparse NSGF-CQT Representations
%   This demo generates a logarithmic chirp, computes its full and sparse
%   nonstationary Gabor frame Constant-Q Transforms (NSGF-CQT),
%   and visualizes the results and their differences.
%   Requires the CiCueTea library and MATLAB's Signal Processing Toolbox.

clear
clc


% Add source code to path
addpath(genpath("../src"));


% Analysis parameters
fs = 48000;           % Sample rate
nSamps = 2^18;        % Number of samples
frac = 1;             % Frequency resolution (periods per octave)
fMin = 100;           % Minimum frequency
fMax = 10000;         % Maximum frequency


% Generate a logarithmic chirp signal and apply a Kaiser window
t = (0:nSamps-1)'/fs;
x = chirp(t, fMin, t(end), fMax, "logarithmic");
w = kaiser(nSamps, 20);
x = x .* w;


% Initialize full and sparse NSGF-CQT
s1 = nsgfCQTInit("full", fs, nSamps, frac);
s2 = nsgfCQTInit("sparse", fs, nSamps, frac);


% Compute full and sparse CQT transforms
X1 = nsgfCQT(x, s1);
X2 = nsgfCQT(x, s2);
X2 = nsgfRasterize(X2, s2); % Convert sparse to matrix form


% Get the time-domain atoms of the full version by inverse real FFT
g1 = s1.g(1:end/2+1,:);
h1 = ifftshift(irfft(g1, s1.nSamps), 1);
h1 = h1 ./ max(abs(h1));

% Get the time-domain atoms of the sparse version by zero-padding and inverse FFT
h2 = zeros(size(h1));
for k = 1:s2.nBands
    g_i = zeros(s2.nSamps, 1);
    g_i(s2.idxs{k}) = s2.g{k};
    g_i = g_i(1:s2.nSamps/2+1);
    h2(:,k) = ifftshift(irfft(g_i, s2.nSamps));
end
h2 = h2 ./ max(abs(h2));


tax = s1.tax; % Time axis
bax = s1.bax; % Frequency axis


% Plot time-frequency representations
figure(1)
clf()
range = [-100, 0];
subplot 311
imagesc(tax, bax, dB(X1).'); % Full transform
colormap jet
colorbar
set(gca, "CLim", range)
title("Full Transform");
set(gca, "YDir", "normal");


subplot 312
imagesc(tax, bax, dB(X2).'); % Rasterized sparse transform
colormap jet
colorbar
set(gca, "CLim", range)
title("Rasterized Sparse Transform");
set(gca, "YDir", "normal");


subplot 313
imagesc(tax, bax, dB(X1-X2).'); % Difference between full and sparse
colormap jet
colorbar
set(gca, "CLim", range)
title("Difference");
set(gca, "YDir", "normal");


% Plot average spectrum and energy over time or frequency
figure(2)
clf();
subplot 211
semilogx(bax, pow2db(mean(abs(X1).^2)), "LineWidth", 2)
hold on
semilogx(bax, pow2db(mean(abs(X2).^2)), "LineWidth", 2)
xlabel("Frequency");
ylabel("Amplitude (dB)");
title("AutoSpectrum");


subplot 212
plot(tax, pow2db(mean(abs(X1).^2, 2)), "LineWidth", 2)
hold on
plot(tax, pow2db(mean(abs(X2).^2, 2)), "LineWidth", 2)
xlabel("Time")
ylabel("Amplitude (dB)")
title("Energy across time")


% Plot time-domain atoms for both full and sparse representations
figure(3)
clf();
subplot 121
plot(tax, h1 + 2 * (1:s1.nBands), "LineWidth", 2);
xlim(0.02 * [-1, 1])
xlabel("Time")
yticks([]);
title("Full time-domain Atoms")


subplot 122
plot(tax, h2 + 2 * (1:s2.nBands), "LineWidth", 2);
xlim(0.02 * [-1, 1])
title("Sparse time-domain Atoms")

File: Python/Demo2.py
import matplotlib.pyplot as plt
import IPython.display as ipd
import numpy as np
import scipy.fft as fft
import scipy.signal as sg
import control as ctl
import NsgfCQT
from Slicing import *
from psychoacoustics import *

def pow2db(x):
    return 10.0 * np.log10(np.abs(x))

def dB(X):
    return 20.0 * np.log10(np.abs(X) + 1e-10)

def rms(x):
    return np.sqrt(np.mean(np.abs(x)**2.0))

fs = 48000
nSamps = 2**18
frac = 1.0
fMin = 100
fMax = 10000

t = np.arange(nSamps) / fs
x = sg.chirp(t, fMin, t[-1], fMax, 'logarithmic')
w = sg.windows.kaiser(nSamps, 20)
x *= w

s1 = NsgfCQT.NsgfCQT('dense', fs, nSamps, frac)
s2 = NsgfCQT.NsgfCQT('sparse', fs, nSamps, frac)

X1 = s1.forward(x)
X2 = s2.forward(x)
X2 = s2.rasterize(X2)

plt.figure(1)
plt.clf()
plt.subplot(3, 1, 1)
plt.pcolormesh(s1.time_axis, s1.band_axis, dB(X1).T)
plt.yscale('log')
plt.clim(-100, 0)

plt.subplot(3, 1, 2)
plt.pcolormesh(s2.time_axis, s2.band_axis, dB(X2).T)
plt.yscale('log')
plt.clim(-100, 0)

plt.subplot(3, 1, 3)
plt.pcolormesh(s2.time_axis, s2.band_axis, dB(X1 - X2).T)
plt.yscale('log')
plt.clim(-100, 0)

g1 = s1.g[:s1.n_freqs//2+1,:]
h1 = fft.irfft(g1, s1.n_samples, 0)
h1 = fft.ifftshift(h1, 0)
h1 = h1 / np.max(np.abs(h1), 0)

h2 = np.zeros_like(h1)
for k in np.arange(s2.n_bands):
    g_i = np.zeros(s2.n_samples)
    g_i[s2.idxs[k]] = s2.g[k]
    g_i = g_i[:s2.n_freqs//2+1]
    h2[:,k] = fft.ifftshift(fft.irfft(g_i, s2.n_samples))
h2 = h2 / np.max(np.abs(h2), 0)

plt.figure(2)
plt.clf()
plt.subplot(2, 1, 1)
plt.semilogx(s1.band_axis, pow2db(np.mean(np.abs(X1)**2.0, 0)))
plt.semilogx(s2.band_axis, pow2db(np.mean(np.abs(X2)**2.0, 0)), "-.")
plt.legend(["dense", "sparse"])

plt.subplot(2, 1, 2)
plt.plot(s1.time_axis, pow2db(np.mean(np.abs(X1)**2.0, 1)))
plt.plot(s2.time_axis, pow2db(np.mean(np.abs(X2)**2.0, 1)), "-.")
plt.legend(["dense", "sparse"])

plt.figure(3)
plt.clf()
plt.subplot(1, 2, 1)
plt.plot(s1.time_axis, h1 + np.arange(s1.n_bands))
plt.xlim([-0.02, 0.02])

plt.subplot(1, 2, 2)
plt.plot(s2.time_axis, h2 + np.arange(s2.n_bands))
plt.xlim([-0.02, 0.02])